### ISSUE TITLE
Prevent IRQ soft mask replay nesting?

### ISSUE THREAD

**Issue by npiggin** [linuxppc/issues#290]
irq_exit in interrupt handlers that get replayed causes irqs to become enabled (local_irq_enable in __do_softirq), which can lead to subsequent interrupt handlers. It's not clear this is any worse than hardware interrupts (which could also fire again when __do_softirq enables interrupts), but this seems to come up more commonly in traces. E.g., from upstream 5.6-rc2 powernv_defconfig

```
swapper     0 [116]   598.431411:      49398 cycles: 
        c000000000c2dd70 radix_tree_lookup+0x10 (/boot/vmlinux)
        c000000000c25600 idr_find+0x20 (/boot/vmlinux)
        c000000000126c8c get_work_pool+0x4c (/boot/vmlinux)
        c000000000129904 __queue_work+0x104 (/boot/vmlinux)
        c000000000129f64 queue_work_on+0x84 (/boot/vmlinux)
        c0000000009dec34 dbs_irq_work+0x34 (/boot/vmlinux)
        c000000000290308 irq_work_run_list+0xa8 (/boot/vmlinux)
        c0000000002903a0 irq_work_run+0x30 (/boot/vmlinux)
        c00000000002e5c8 timer_interrupt+0x208 (/boot/vmlinux)
        c00000000000a93c replay_interrupt_return+0x0 (/boot/vmlinux)
        c00000000001cc44 arch_local_irq_restore.part.8+0xc4 (/boot/vmlinux)
        c000000000c47cb0 _raw_spin_unlock_irqrestore+0x50 (/boot/vmlinux)
        c000000000147cf8 try_to_wake_up+0x98 (/boot/vmlinux)
        c0000000001b5304 process_timeout+0x24 (/boot/vmlinux)
        c0000000001b5370 call_timer_fn+0x50 (/boot/vmlinux)
        c0000000001b6110 run_timer_softirq+0x750 (/boot/vmlinux)
        c000000000c486f8 __do_softirq+0x158 (/boot/vmlinux)
        c0000000001071b4 irq_exit+0xf4 (/boot/vmlinux)
        c00000000002e4e8 timer_interrupt+0x128 (/boot/vmlinux)
        c00000000000a93c replay_interrupt_return+0x0 (/boot/vmlinux)
        c00000000001cc44 arch_local_irq_restore.part.8+0xc4 (/boot/vmlinux)
        c0000000009e2070 cpuidle_enter_state+0xf0 (/boot/vmlinux)
        c0000000009e2630 cpuidle_enter+0x50 (/boot/vmlinux)
        c00000000014e30c call_cpuidle+0x4c (/boot/vmlinux)
        c00000000014e908 do_idle+0x2e8 (/boot/vmlinux)
        c00000000014ec38 cpu_startup_entry+0x38 (/boot/vmlinux)
        c0000000000512bc start_secondary+0x64c (/boot/vmlinux)
        c00000000000ac54 start_secondary_prolog+0x10 (/boot/vmlinux)
```

Should go through and ensure irq_entry/irq_exit match what other architectures are doing, and see if there is a way to reduce or avoid this nesting coming up for replay.

**Comment by npiggin** [comment]
Possible solution here

https://lists.ozlabs.org/pipermail/linuxppc-dev/2022-October/249496.html

On one hand the existing nested replay is basically matching what hardware IRQs can do on powerpc and other architectures, so maybe(?) it's okay to leave as is. On the other hand, the software re-entrancy has more complexity than simply new hardware interrupts coming in asynchronously.

**Comment by mpe** [comment]
I have always disliked the nesting, it confuses people when it appears in backtraces, as it looks like we're recursing.

I was trying to document how the nesting happens recently for someone, and came up with the below, which I think is correct. Mainly though it just demonstrates how long and involved the path is. Would be nice if we could simplifiy it.

```
local_irq_disable()
...
-> hard interrupt
   exception entry sets paca->irq_happened
   rfi
...
local_irq_enable()
  raw_local_irq_enable()
    arch_local_irq_enable()
      arch_local_irq_restore(IRQS_ENABLED);
        if irq_happened:
          __hard_irq_disable()
          replay_soft_interrupts()
            timer_interrupt()   // for example
              if should_hard_irq_enable():
                do_hard_irq_enable()
                  __hard_irq_enable()   // NB interrupts remain soft disabled
              ...
              interrupt_async_exit_prepare()
                irq_exit()
                  __irq_exit_rcu()
                    invoke_softirq()
                      do_softirq_own_stack()
                        call_do_softirq()
                          __do_softirq()
                            local_irq_enable()
                              raw_local_irq_enable()
                                arch_local_irq_enable()
                                  arch_local_irq_restore(IRQS_ENABLED);
                                    if irq_happened:
                                      __hard_irq_disable()
                                      replay_soft_interrupts()
                                        timer_interrupt()   // for example
                                        ...
                                        irq_exit()
                                          __irq_exit_rcu()
                                            if (in_interrupt())
                                              return
```

**Comment by npiggin** [comment]
Yeah that looks about right. We can nest even if should_hard_irq_enable() == false because __do_softirq() does a local_irq_enable(); ... local_irq_disable(); so we end up being soft-masked and can get new pending irqs come in there. So it wouldn't be enough to just do the PACA_IRQ_REPLAYING from the above patch to suppress should_hard_irq_enable().

**Comment by npiggin** [comment]
Fixed with 5746ca131e24 ("powerpc/64: Don't recurse irq replay")
